### ISSUE TITLE
Preview / Query revisions

### ISSUE THREAD

**Issue by heggemsnes** [wp-graphql/wp-graphql-woocommerce#421]
Hi!

I'm setting up a woocommerce website with Next.js front end. I've been using this base project: https://github.com/vercel/next.js/tree/canary/examples/cms-wordpress and have set up so previews on posts works fine.

It doesn't for products though. I've added revisions to Products through functions.php but I'm not able to query revisions on the product.

`// Allow revisions on Products
add_filter( 'woocommerce_register_post_type_product', 'wc_modify_product_post_type' );

function wc_modify_product_post_type( $args ) {
     $args['supports'][] = 'revisions';

     return $args;
}`

Does anyone have any experience with previewing products and/or know how to add Product revisions to graphql?

**Comment by ardiewen** [comment]
hey @heggemsnes , there seems to be a gap in functionality here.

On the `wp-graphql` side of things, there is an argument flag `asPreview` that works for all post types, including custom post types, that support revisions.

Since products are just another custom post type, this should "just work", but it looks like wp-graphql  / wp-graphql-woocommerce treats products differently.

@kidunot89, @jasonbahl  is this something that we can get implemented, such that wp-graphql-woocommerce supports the same preview functionality as the base wp-graphql package?

Here is wp-graphql's implementation of the argument for [allowed post objects](https://github.com/wp-graphql/wp-graphql/blob/d4f9fb560fa6bb8d8ce90ecd0090568de39640dd/src/Type/Object/RootQuery.php#L490-L492)

Here is wp-graphql's implementation of [wp_get_post_revisions](https://github.com/wp-graphql/wp-graphql/blob/d4f9fb560fa6bb8d8ce90ecd0090568de39640dd/src/Type/Object/RootQuery.php#L525-L532) to replace the post_id if the argument is true

And finally, here is wp-graphql's implementation for the [product root query](https://github.com/wp-graphql/wp-graphql-woocommerce/blob/bc1c826e7c462fb56ae370c9e4afcd4af272e177/includes/type/interface/class-product.php#L52-L106)

For what its worth, I forked this repo and tried adding the lines for the `asPreview` argument, as well as the function below to run `wp_get_post_revisions` and modify $product_id accordingly, and then I got stuck.

Basically, the revision_id that comes back doesn't carry the post_type="Product" and throws an error.

Happy to assist further if I can, but I'm a bit beyond my element here.

**Comment by kidunot89** [comment]
@heggemsnes as far as I know the Product and Product Variation post types don't support revisions 🤔.

**Comment by ardiewen** [comment]
> @heggemsnes as far as I know the Product and Product Variation post types don't support revisions 🤔.

With the filter mentioned above, you can enable support insofar as the post content. E.g. prices and product metadata **will not** be stored, but name, description, and ACF fields should be in scope.

If this is the wrong approach, perhaps another worth considering would be to:

1. Create a CPT called `Product Content` or something similar.

2. Use something like ACF to create a post-object relationship from the CPT to the actual product.

**Comment by kidunot89** [comment]
@ardiewen In order for revisions to work for a custom post type like `product` and you need to [enable it](https://wordpress.stackexchange.com/questions/139031/enable-revisions-for-custom-post-type) on the WordPress level not just in WooGraphQL. By design WooCommerce doesn't add revision support for [`product`](https://github.com/woocommerce/woocommerce/blob/master/includes/class-wc-post-types.php#L288). This can be easily remedied with filters like you said, but I personally don't know if making that change will break something on WooCommerce product edit page. :man_shrugging:

**Comment by ardiewen** [comment]
@kidunot89 I agree with this direction, and personally have moved to create a CPT for my product marketing data instead of having it on the actual product post type.

However, there still is an issue with Gatsby Incremental Builds. The combination of:

1. the product post type doesn't support revisions by default
2. gatsby triggers an incremental build when a product is updated and errors out because it can't find the `asPreview` argument.

This feels like more of an issue on the source plugin / wp-gatsby side, so I'll look to address it in an issue there.
